crates/wasmparser/src/validator/operators.rs

```diff
@@ -3442,6 +3473,79 @@ where
         self.pop_operand(Some(ValType::I32))?;
         Ok(())
     }
+    fn visit_struct_new_default(&mut self, type_index: u32) -> Self::Output {
+        let ty = self.struct_type_at(type_index)?;
+        for i in 0..u32::try_from(ty.len_fields()).unwrap() {
+            let field = ty.field_at(i).unwrap();
+            let val_ty = field.element_type.unpack();
+            if !val_ty.is_defaultable() {
+                bail!(
+                    self.offset,
+                    "invalid `struct.new_default`: {val_ty} field is not defaultable"
+                );
+            }
+        }
+
+        let mut heap_ty = HeapType::Concrete(UnpackedIndex::Module(type_index));
+        self.resources.check_heap_type(&mut heap_ty, self.offset)?;
+
+        let ref_ty = RefType::new(false, heap_ty).ok_or_else(|| {
+            format_err!(self.offset, "implementation limit: type index too large")
+        })?;
+
+        self.push_operand(ref_ty)
+    }
+    fn visit_array_new_default(&mut self, type_index: u32) -> Self::Output {
+        let ty = self.array_type_at(type_index)?;
+        let field = ty.field_type();
+        let val_ty = field.element_type.unpack();
+        if !val_ty.is_defaultable() {
+            bail!(
+                self.offset,
+                "invalid `array.new_default`: {val_ty} field is not defaultable"
+            );
+        }
+
+        let mut heap_ty = HeapType::Concrete(UnpackedIndex::Module(type_index));
+        self.resources.check_heap_type(&mut heap_ty, self.offset)?;
+
+        let ref_ty = RefType::new(false, heap_ty).ok_or_else(|| {
+            format_err!(self.offset, "implementation limit: type index too large")
+        })?;
+
+        self.pop_operand(Some(ValType::I32))?;
+        self.push_operand(ref_ty)
+    }
+    fn visit_any_convert_extern(&mut self) -> Self::Output {
+        let extern_ref = match self.pop_ref()? {
+            Some(r) if self.resources.top_type(&r.heap_type()) == HeapType::Extern => r,
+            Some(r) => bail!(
+                self.offset,
+                "type mismatch: expected (ref null? extern), found {r}"
+            ),
+            None => bail!(
+                self.offset,
+                "type mismatch: expected (ref null? extern), found bottom"
+            ),
+        };
```

Comment:
I'm a little surprised that subtyping checks didn't kick in here, could this not be `self.pop_operand` with `RefType::EXTERNREF` or similar?

Reply:
(same below for popping `anyref`)

Reply:
This does take subtyping into account: when checking if `t <: Any` we can just do `top(t) == Any`, which is going to be faster.

We can't use `self.pop_operand(Some(RefType::EXTERNREF))` because these instructions are polymorphic over nullability. We don't know ahead of time whether to expect a nullable or non-nullable reference.

Reply:
Oh sorry I mean moreso for I would have expected that `self.pop_operand` to work since it has built-in subtyping checks. I'm not sure what you mean though precisely about the instructions being polymorphic? If it's what I think though `pop_operand` returns the precise type which I think can be tested for being nullable or non-nullable?

Reply:
Its that we can't pass in an expected type to pop because we can pop either a `(ref extern)` or a `(ref null extern)`

Reply:
Isn't `(ref extern)` a subtype of `(ref null extern)` though so you could pop a `(ref null extern)` and see what came out?

Reply:
Ah now I see what you are getting at. Pushed a new commit that does this.